File: content/problems/1610-maximum-number-of-visible-points/solutions/circular-angle-window/Solution.java
class Solution {
    public int visiblePoints(List<List<Integer>> points,int angle,List<Integer> location){
        if(angle>=360)return points.size();
        List<Double> angles=new ArrayList<>();
        int same=0;
        for(List<Integer> point:points){
            int dx=point.get(0)-location.get(0),dy=point.get(1)-location.get(1);
            if(dx==0&&dy==0)same++;
            else angles.add(Math.atan2(dy,dx));
        }
        Collections.sort(angles);
        int original=angles.size();
        for(int i=0;i<original;i++)angles.add(angles.get(i)+2*Math.PI);
        double width=Math.toRadians(angle);
        int left=0,best=0;
        for(int right=0;right<angles.size();right++){
            while(angles.get(right)-angles.get(left)>width+1e-12)left++;
            best=Math.max(best,Math.min(original,right-left+1));
        }
        return best+same;
    }
}

File: content/problems/1610-maximum-number-of-visible-points/solutions/circular-angle-window/Solution.kt
class Solution {
    fun visiblePoints(points:List<List<Int>>,angle:Int,location:List<Int>):Int{
        if(angle>=360)return points.size;
        val angles=mutableListOf<Double>();
        var same=0;
        for(point in points){
            val dx=point[0]-location[0];
            val dy=point[1]-location[1];
            if(dx==0&&dy==0)same++ else angles.add(kotlin.math.atan2(dy.toDouble(),dx.toDouble()))
        };
        angles.sort();
        val original=angles.size;
        for(i in 0 until original)angles.add(angles[i]+2*Math.PI);
        val width=Math.toRadians(angle.toDouble());
        var left=0;
        var best=0;
        for(right in angles.indices){
            while(angles[right]-angles[left]>width+1e-12)left++;
            best=maxOf(best,minOf(original,right-left+1))
        };
        return best+same
    }
}
